pkg/health/condition/check_backup_ready.go

```diff
@@ -37,101 +38,184 @@ const (
+	lease := &coordinationv1.Lease{}
+	err := a.cl.Get(ctx, types.NamespacedName{Name: name, Namespace: namespace}, lease)
+	return lease, err
+}
 
-		// Transition to "False" state only if present state is "Unknown" or "False"
-		if deltaLeaseRenewTime != nil && (prevBackupReadyStatus.Status == druidv1alpha1.ConditionUnknown || prevBackupReadyStatus.Status == druidv1alpha1.ConditionFalse) {
-			if time.Since(deltaLeaseRenewTime.Time) > 3*etcd.Spec.Backup.DeltaSnapshotPeriod.Duration {
-				result.status = druidv1alpha1.ConditionFalse
-				result.reason = BackupFailed
-				result.message = "Stale snapshot leases. Not renewed in a long time"
-				return result
-			}
+func createBackupConditionResult(status druidv1alpha1.ConditionStatus, reason string, message string) *result {
+	return &result{
+		conType: druidv1alpha1.ConditionTypeBackupReady,
+		status:  status,
+		reason:  reason,
+		message: message,
+	}
+}
+
+func isLeaseStale(renewTime time.Time, renewalGracePeriod time.Duration) bool {
+	return time.Since(renewTime) > renewalGracePeriod
+}
+
+func wasLeaseCreatedRecently(creationTime time.Time, creationGracePeriod time.Duration) bool {
+	return time.Since(creationTime) < creationGracePeriod
+}
+
+// handleOnlyDeltaSnapshotLeaseRenewedCase handles cases where only delta snapshot lease is renewed,
+// while the full snapshot lease is not renewed, possibly since it might have been recreated recently.
+// Treats backup as succeeded if delta snapshot lease is renewed within the required time window
+// and full snapshot lease object is not older than the computed full snapshot duration.
+func handleOnlyDeltaSnapshotLeaseRenewedCase(fullSnapshotLeaseCreationTime time.Time, fullSnapshotLeaseCreationGracePeriod time.Duration, deltaSnapshotLeaseRenewTime time.Time, deltaSnapshotLeaseRenewalGracePeriod time.Duration) *result {
+	wasFullSnapshotLeaseCreatedRecently := wasLeaseCreatedRecently(fullSnapshotLeaseCreationTime, fullSnapshotLeaseCreationGracePeriod)
+	isDeltaSnapshotLeaseStale := isLeaseStale(deltaSnapshotLeaseRenewTime, deltaSnapshotLeaseRenewalGracePeriod)
+
+	if isDeltaSnapshotLeaseStale {
+		if wasFullSnapshotLeaseCreatedRecently {
+			return createBackupConditionResult(
+				druidv1alpha1.ConditionFalse, BackupFailed,
+				"Delta snapshot backup failed. Delta snapshot lease not renewed in a long time",
+			)
+		} else {
+			return createBackupConditionResult(
+				druidv1alpha1.ConditionFalse, BackupFailed,
+				"Stale snapshot leases. Not renewed in a long time",
+			)
 		}
 	}
 
-	//Transition to "Unknown" state is we cannot prove a "True" state
-	return result
+	if !wasFullSnapshotLeaseCreatedRecently {
+		return createBackupConditionResult(
+			druidv1alpha1.ConditionFalse, BackupFailed,
+			"Full snapshot backup failed. Full snapshot lease created long ago, but not renewed",
+		)
+	}
+
+	return createBackupConditionResult(
+		druidv1alpha1.ConditionTrue, BackupSucceeded,
+		"Delta snapshot backup succeeded",
+	)
```

Comment:
@Shreyas-s-rao,

Nitpick: Rather than using nested conditions, what do you think about flattening them into a switch case structure as follows?

```go
	switch {
	case !deltaSnapshotLeaseRenewedRecently && fullSnapshotLeaseCreatedRecently:
		// Delta snapshot lease is stale, but full snapshot lease was created recently
		return createBackupConditionResult(druidv1alpha1.ConditionFalse, BackupFailed, "Delta snapshot backup failed. Delta snapshot lease not renewed in a long time")
	case !deltaSnapshotLeaseRenewedRecently && !fullSnapshotLeaseCreatedRecently:
		// Both leases are stale
		return createBackupConditionResult(druidv1alpha1.ConditionFalse, BackupFailed, "Stale snapshot leases. Not renewed in a long time")
	case !fullSnapshotLeaseCreatedRecently:
		// Full snapshot lease is stale, but delta snapshot lease was renewed recently
		return createBackupConditionResult(druidv1alpha1.ConditionFalse, BackupFailed, "Full snapshot backup failed. Full snapshot lease created long ago, but not renewed")
	default:
		// Delta snapshot lease was renewed recently and full snapshot lease is not stale
		return createBackupConditionResult(druidv1alpha1.ConditionTrue, BackupSucceeded, "Delta snapshot backup succeeded")
	}
```

Alternatively, for consistency, we could consider maintaining similar flattened conditions as those in the `handleRenewedSnapshotLeases` function. [Link to Code](https://github.com/gardener/etcd-druid/pull/619/files#diff-9cc4496b69f9d0fb63a6e6fff5162bfbccae4e70edf42fff483f9da523751ca2R194-R218)

Reply:
Problem with switch cases is that even though we specify a `default` case at the end which will surely be executed if none of the previous cases are satisfied, golang expects a return statement at the end of the function in order to keep things absolutely deterministic. So we will be forced to either remove the default case from the switch construct and move it out to the end of the function, or to return a dummy backupReady result, which is semantically incorrect.

I would like to retain the if-else construct way of handling the different cases, but I'll flatten it out as you requested.

Reply:
Thanks Shreyas